### PR TITLE
JBEAP-7598 Fix memory leak in JMX protocol

### PR DESCRIPTION
**Pull request by ctomc**
Also cleanup how arquillian services are created and use injection instead of params in constructor.

### CODE REVIEW

**Comment by jamezp** [comment]
I need to look at this in more detail, but for a 2.0.x release we definitely can't have the first commit. Would you mind explaining a bit on why the first commit is required? If it's just some clean up that's fine, we just might need to hold off is all.

**Comment by ctomc** [comment]
Why can't we **definitely** have first commit?
Code changes how references to deployment are done, which in results in easier GC as we are using injected values for DeploymentUnit instead of manual lookup and passing in constructor.
Also logic for creating and maintaining lifecycle of ArquillianConfig
In my testing it also shows that it is also somewhat faster with this change.
it should be tested bit more for sure, but I don't see why we cannot merge it.

If that means 2.1.x will be release version, sure, let that be it. As long as we get release that we can use for testing WildFly. As this changes only affect wildfly/eap as we are the only users of JMX protocol.

**Comment by jamezp** [comment]
The main reason is the `ArquillianConfigBuilder` change from public to package-private. While I agree package-private is the correct scope, we just can't do it in a macro release IMO. I'll look in more detail at the other changes though.

**Comment by ctomc** [comment]
package-private is only concern I don't mind reverting that

**Comment by jamezp** [comment]
The changes seem okay. I kind of wonder if we should just redo a bunch of the service stuff from scratch though. It all looks, as I'm sure you saw :), a bit messy. For example we should look at removing the `ServiceListener` since it's deprecated.

What do you think? Obviously this would have to be in a 2.1 release, but we may need to bump the core version anyway for Elytron so maybe we just make master 2.1 anyway and make a 2.0.x branch.

**Comment by jamezp** [comment]
BTW I'm not suggesting you do a rewrite, unless you want to, just getting your opinion on it :) I can try to find some time to do it if you also think it's a good idea.

**Comment by ctomc** [comment]
Yeah, this PR also does first step in this direction, it changes the way how whole services are crated to be much ,much cleaner. 
It is first step, we need to improve also that.
could you merge this and do 2.1.Beta1 release or something like that.
I could take a look at how to remove service listener as well.

File: protocol-jmx/src/main/java/org/jboss/as/arquillian/service/ArquillianService.java [hunk 1]
```diff
@@ -64,73 +62,31 @@
...
 
     private final InjectedValue<MBeanServer> injectedMBeanServer = new InjectedValue<MBeanServer>();
-    private final Set<ArquillianConfig> deployedTests = new HashSet<ArquillianConfig>();
-    private ServiceContainer serviceContainer;
-    private ServiceTarget serviceTarget;
-    private JMXTestRunner jmxTestRunner;
-    AbstractServiceListener<Object> listener;
+    private final Set<ArquillianConfig> deployedTests = new HashSet<>();
+    private volatile JMXTestRunner jmxTestRunner;
+    private volatile AbstractServiceListener<Object> listener;
```
Comment (jamezp) [inline protocol-jmx/src/main/java/org/jboss/as/arquillian/service/ArquillianService.java]: I won't hold up the PR for this, but there is no reason for these to be `volatile`. The `start()` and `stop()` methods are both synchronized.

**Pull request merged**